jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/SingleThreadedLogger.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2018 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.plugins.common;
+
+import com.google.common.util.concurrent.Futures;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.function.Consumer;
+
+/**
+ * Keeps all log messages in a sequential, deterministic order along with an additional footer that
+ * always appears below log messages.
+ */
+class SingleThreadedLogger {
+
+  /** ANSI escape sequence for moving the cursor up one line. */
+  private static final String CURSOR_UP_SEQUENCE = "\033[1A";
+
+  private final ExecutorService executorService = Executors.newSingleThreadExecutor();
+
+  private final Consumer<String> plainLogger;
+
+  private String footer = "";
+  private int footerLineCount = 0;
+
+  SingleThreadedLogger(Consumer<String> plainLogger) {
+    this.plainLogger = plainLogger;
+  }
+
+  /**
+   * Runs {@code messageLogger} asynchronously.
+   *
+   * @param messageLogger the {@link Runnable} intended to synchronously log a message to the
+   *     console
+   * @return a {@link Future} to track completion
+   */
+  public Future<Void> log(Runnable messageLogger) {
```

Comment:
Trying to understand why this needs to be a `Runnable` instead of a `String`. Isn't it that in the end what this `Runnable` will do is to write some string to `plainLogger`? Actually, what does "plain" mean here? Also, is it that this "message logger" (who will log a message to the console according to the Javadoc) is a different logger than the plain logger? (I didn't get what these two loggers will be in the end.)

Reply:
So this is for the user to pass like (Maven) `() -> log.info("log message")` or other logging method, so that this class is independent of the actual different logging levels.

`plain` is for a logging method similar to just plain `System.out.print`, so it will be like `lifecycle` for Gradle and probably `System.out.print` for Maven (might need to do some things differently if wanting to use Maven info logging).

Reply:
Ah, now I see the intention. Basically, this class is the `LogEvent` handler (or will be used by whatever `LogEvent` handler) to "move" the footer (the multi-line progress display) down by one line after logging the message (if it gets printed to the screen, for example). This is done by calling this `log()` message. And at the same time, this class is the `ProgressEvent` handler (or will be used by whatever `ProgressEvent` handler`) to update the progress display, which is done by `setFooter()`. I see where my confusion came from: first, the potential mix of `LogEvent` and `ProgressEvent`; secondly, the mix of two output channels, one for `messageLogger` and `plainLogger` (for example, one going out to a file and the other going out to a screen); and the `messageLogger` is basically an unrestricted lambda. Not that I am suggesting something concrete here... don't really have an idea.

Reply:
Yea, this will be used by the `LogEventHandler`. A `ProgressLogger` will be the one that handles `ProgressEvent`s and directly calls `setFooter` on this or set the footer via a `LogEvent` type of `CHANGE_FOOTER` or something.